fmpz/multi_crt.c

```diff
@@ -362,78 +354,177 @@ int fmpz_multi_crt(
     fmpz_swap(out + 0, output);
 
     for (i = 0; i < P->localsize; i++)
-    {
         fmpz_clear(out + i);
-    }
 
-    fmpz_multi_crt_clear(P);
+    fmpz_multi_CRT_clear(P);
 
     TMP_END;
 
     return success;
 }
 
 
-/*
-    If P was set with a call to fmpz_poly_crt_compile(P, m, len), return
-    in outputs[0] signed integer r of smallest abs value such that
-        r = inputs[0] mod m[0]
-        r = inputs[1] mod m[1]
-            ...
-        r = inputs[len-1] mod m[len-1]
-    For thread safety "outputs" is expected to have enough space for all
-    temporaries, thus should be at least as long as P->localsize.
-*/
-
-void _fmpz_multi_crt_run(
+void _fmpz_multi_CRT_run(
     fmpz * outputs,
-    const fmpz_multi_crt_t P,
-    const fmpz * inputs)
+    const fmpz_multi_CRT_t P,
+    const fmpz * inputs,
+    int sign)
 {
-    slong i;
-    slong a, b, c;
-    const fmpz * B, * C;
-    fmpz * A, * t1, * t2;
+    slong i, a, b, c;
+    slong len = P->length;
+    const fmpz * m = P->moduli;
+    const fmpz * mf = P->fracmoduli;
+    fmpz * A, * B, * C, * t1, * t2, * t3, * t4;
 
     t1 = outputs + P->temp1loc;
     t2 = outputs + P->temp2loc;
+    t3 = outputs + P->temp3loc;
+    t4 = outputs + P->temp4loc;
 
-    for (i = 0; i < P->length; i++)
+    FLINT_ASSERT(len < 1 || P->good);
+
+    if (len > 0)
+    {
+        for (i = P->moduli_count - 1; i > 0; i--)
+        {
+            if (!fmpz_equal(inputs + 0, inputs + i))
+                goto doit;
+        }
+    }
+
+    _fmpz_mods(outputs + 0, inputs + 0, P->final_modulus, sign, t4);
+    return;
+
+doit:
+
+    for (i = 0; i < len; i++)
     {
         a = P->prog[i].a_idx;
         b = P->prog[i].b_idx;
         c = P->prog[i].c_idx;
-        FLINT_ASSERT(a >= 0);
+
         A = outputs + a;
-        B = b < 0 ? inputs + (-b-1) : outputs + b;
-        C = c < 0 ? inputs + (-c-1) : outputs + c;
+        B = outputs + b;
+        C = outputs + c;
+
+        if (b < 0)
+        {
+            b = -b - 1;
+            B = t1;
 
-        /* A = B + I*(C - B) mod M */
-        fmpz_sub(t1, B, C);
-        fmpz_mul(t2, P->prog[i].idem, t1);
-        fmpz_sub(t1, B, t2);
-        fmpz_smod(A, t1, P->prog[i].modulus);
+            fmpz_mul(t3, inputs + b, mf + b);
+            _fmpz_mods(B, t3, m + b, sign, t4);
+        }
 
-        /* last calculation should write answer to outputs[0] */
-        if (i + 1 >= P->length)
+        if (c < 0)
         {
-            FLINT_ASSERT(A == outputs + 0);
+            c = -c - 1;
+            C = t2;
+
+            fmpz_mul(t3, inputs + c, mf + c);
+            _fmpz_mods(C, t3, m + c, sign, t4);
         }
+
+        /* A = B*c_m + C*b_m */
+        fmpz_mul(A, B, P->prog[i].c_modulus);
+        fmpz_mul(t3, C, P->prog[i].b_modulus);
+        fmpz_add(A, A, t3);
     }
+
+    _fmpz_mods(outputs + 0, A, P->final_modulus, sign, t4);
+}
+
+
+/* depreciated functions *****************************************************/
```

Comment:
deprecated

Reply:
Maybe make a note in the docs about deprecation. Also mention any downsides of calling the deprecated function, e.g. slower.

Reply:
Please also send me a full list of deprecated functions so I can mention in the release notes.